API changes for listing/marking source conversation items that have been seen by journalists

rmol opened the pull request:
## Status

Ready for review

## Description of Changes
    
Add new API endpoint for listing or marking source conversation items that have been seen by a journalist.
    
Add utility method to mark a heterogeneous collection of Submission and Reply objects seen.
    
Add Submission.is_file and Submission.is_message to encapsulate the characterization based on filename.

Fixes #5475.

## Testing

- Run the dev server with `make dev`.

- Get an API token as the `journalist` user:
  ```
  export TOKEN=$(curl -o- -X POST -H "Content-Type: application/json" --data '{"username":"journalist","passphrase":"correct horse battery staple profanity oil chewy","one_time_code":"047889"}' http://localhost:8081/api/v1/token | jq -r .token)
  ```

- List current submissions: 
  ```
  curl -X GET -H "Content-Type: application/json" -H "Authorization: Token $TOKEN" http://localhost:8081/api/v1/submissions
  ```
  All should have a `seen_by` property, currently an empty list.

- Mark a message seen:
  ```
  curl -X POST -H "Content-Type: application/json" -H "Authorization: Token $TOKEN" --data '{"messages": ["e711d29d-d0a1-440f-80e4-6642b77ec3ea"]}' http://localhost:8081/api/v1/seen
  ```
  Replace the UUID with one from one of the messages in the response from the `/submissions` endpoint, of course.
- Mark the same message seen again. There should be no error.
- Retrieve `/submissions` again. The message's `seen_by` list should now contain your journalist account's UUID.
- Visit the source of that message in the journalist interface. The message you marked should not be listed in bold text. Clicking "Select unread" should not select that message.
- Visit the source list in the journalist interface. Select the source of the message you marked and click "Download Unread". The zip file you receive should not contain the seen message.

- Via the source interface, upload a file, then retrieve `/submissions` again, and mark the file seen:
  ```
  curl -X POST -H "Content-Type: application/json" -H "Authorization: Token $TOKEN" --data '{"files": ["4fe4b7f5-3031-4651-9092-888951681cf7"]}' http://localhost:8081/api/v1/seen
  ```
  Replace the UUID with that of the file you just uploaded, of course.
- Repeat the checks in the journalist interface that you performed for a seen message, confirming that the file you marked seen is always considered read.

- Get another API token, this time as the `dellsberg` journalist account.
- Retrieve the list of all replies:
  ```
  curl -X GET -H "Content-Type: application/json" -H "Authorization: Token $TOKEN" http://localhost:8081/api/v1/replies
  ```
- Mark one of them read: 
  ```
  curl -X POST -H "Content-Type: application/json" -H "Authorization: Token $TOKEN" --data '{"replies": ["8924a691-3b0f-45c1-bfed-b0b91e264855"]}' http://localhost:8081/api/v1/seen
  ```
- Hit `/replies` again and confirm that the `dellsberg` account UUID appears in the reply's `seen_by` list.

## Deployment

Depends on the database changes in #5505.

## Checklist

### If you made changes to the server application code:

- [ ] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

Commit:
store seen-by when files are downloaded

sssoleileraaa commented on securedrop/create-dev-data.py:
```diff
@@ -138,6 +138,9 @@ def create_source_and_submissions(
         reply = Reply(journalist, source, fname)
         db.session.add(reply)
+        db.session.flush()
+        seen_reply = SeenReply(reply_id=reply.id, journalist_id=journalist.id)
```
nice change because if you were to do this from the JI this would happen automatically

sssoleileraaa commented on securedrop/journalist_app/api.py:
```diff
@@ -309,6 +314,57 @@ def get_all_replies():
+        """
+        user = get_user_object(request)
+
+        if request.method == 'GET':
```
We discussed in [the spec](https://docs.google.com/document/d/1szPEKI1gTpRiM_YI87plKgiW39T7zT4Y4HG0uFnsJD4)two possible implementations for this so was wondering why you went in this direction (perhaps for performance reasons as described in https://github.com/freedomofpress/securedrop/issues/5475?)

rmol replied:
Both are implemented here (though I just realized `/seen`'s `GET` is incomplete, as it's only returning the authenticated client's seen items -- will fix when I rebase to catch up to the latest version of #5505).

sssoleileraaa replied:
okay, I see that you added seen records in the /replies and /submissions endpoints :+1:. why did you also add a /seen endpoint to get all seen records for the currently logged in journalist. We never discussed doing this since it's not in scope for the seen/unseen feature - we did discuss providing **all** seen records in a new endpoint as an alternative option to updating the /replies and /submissions endpoint. or am i misunderstanding your change? it could be useful to start a docs draft pr similar to https://github.com/freedomofpress/securedrop-docs/pull/3 to provide clarity for reviewers.

rmol replied:
In the comment immediately above this, I mentioned that it was incorrect at that point. It has since been changed to match the spec, returning all the records. If you're reviewing locally, make sure to pull frequently. Or just wait; I should be moving this from draft to ready for review shortly.

sssoleileraaa replied:
oops sorry about that. will wait!

sssoleileraaa reviewed (commented):
did a first pass

sssoleileraaa commented:
Not sure how the API is going to look once you're done, but just as a reminder, we were also considering updating the current replies and submissions endpoints as an option:

>[Option 2] Update current endpoints

```yml
"replies": [
        {
            "filename": <filename>,
            ...
            "seen_by": [<journalist_uuid>, ]
        },
        {
            "filename": <filename>,
            ...
            "seen_by": [<journalist_uuid>, ]
        },
]

"submissions": [
        {
            "is_read": <is_read>,
            "filename": <filename>,
            ...
            "seen_by": [<journalist_uuid>, ]
            "opened_by": [<journalist_uuid>, ]
        },
        {
            "is_read": <is_read>,
            "filename": <filename>,
            ...
            "seen_by": [<journalist_uuid>, ]
            "opened_by": [<journalist_uuid>, ]
        },
]
```

sssoleileraaa commented:
just a thought, we could rename "is_read" to "downloaded_from_web" ~or "seen_from_web"~

rmol commented:
> Not sure how the API is going to look once you're done, but just as a reminder, we were also considering updating the current replies and submissions endpoints as an option

That is already present, as suggested in the test plan. Look at the `to_json` methods of [`Submission`](https://github.com/freedomofpress/securedrop/blob/078a022392fd7ce423345cdcf12fd507c38ae47b/securedrop/models.py#L307) and [`Reply`](https://github.com/freedomofpress/securedrop/blob/078a022392fd7ce423345cdcf12fd507c38ae47b/securedrop/models.py#L379) in `models.py`.

rmol commented:
> just a thought, we could rename "is_read" to "downloaded_from_web"

I know we've been playing pretty fast and loose with the API and versioning (at least it's mostly been additions :neutral_face: ), but `is_read` is in production. It's possible that someone is using the API for something other than the workstation, so we should at least pretend we care about API stability.

rmol committed:
Add /seen API endpoint, utils.mark_seen, submission type props

Add new API endpoint for listing or marking source conversation items
that have been seen by a journalist.

Add utility method to mark a heterogeneous collection of Submission
and Reply objects seen.

Add Submission.is_file and Submission.is_message to encapsulate the
characterization based on filename.

rmol committed:
Drop GET handling at /seen API endpoint

rmol committed:
Mark submissions and replies seen the same way in more places

With the addition of journalist_app.utils.mark_seen, it can replace
the helper methods in tests/utils/db_helper.py, and the ad hoc logic
to mark things seen in journalist_app/col.py.

sssoleileraaa commented on securedrop/models.py:
```diff
@@ -212,7 +211,23 @@ def __init__(self, source: Source, filename: str) -> None:
 
-    def to_json(self) -> 'Dict[str, Union[str, int, bool]]':
+    @property
+    def is_file(self) -> bool:
```
thought about doing this myself in #5505 but the pr was already so large. nice change.

sssoleileraaa commented on securedrop/models.py:
```diff
@@ -212,7 +211,23 @@ def __init__(self, source: Source, filename: str) -> None:
+    def to_json(self) -> "Dict[str, Union[str, int, bool]]":
+        seen_by = {
+            f.journalist.uuid for f in SeenFile.query.filter(SeenFile.file_id == self.id)
+            if f.journalist
```
i see, so this won't include the fact that there is a deleted journalist who saw a file. gonna keep reviewing and will cycle back to this. could be an issue because we still send "deleted" uuids for replies when journalists are deleted. this takes a different course.

rmol replied:
It does not. I didn't see any way for sensible inclusion of those records in `seen_by` here. We do have `Submission.seen`, which will reflect those, and until we start keeping deleted journalists around, we can't distinguish between the records for the submission in the seen tables anyway, so since the only purpose of those rows at this point is establishing the global seen state, I thought `Submission.seen` the right answer.

sssoleileraaa replied:
There are also replies we have to worry about which don't have a global seen state, but I definitely see why you made this choice. Agreed that implementing https://github.com/freedomofpress/securedrop/issues/5467 + global user account for data migration will make this work fine.

rmol replied:
Replies do have a global seen state: if one exists, its creator has seen it. `SeenReply` can tell us who else has seen it, but only if  `journalist_id` is not null, so again, I didn't see any point in including records where it is.

sssoleileraaa reviewed (commented):
first pass through the code - going to start manually testing

conorsch reviewed (approved):
Detailed test plan checks out, functionality matches expectations. Visual review of the diff was smooth, as well. As discussion during review indicates, we should queue up some "v2" refinements for the API.
